xds/internal/balancer/cdsbalancer/cluster_handler_test.go

```diff
@@ -0,0 +1,609 @@
+			}
+
+			// Above represents same thing as the simple TestSuccessCaseLeafNode, extra behavior + validation (clusterNode
+			// which is a leaf receives a changed clusterUpdate, which should ping clusterHandler, which should then write
+			// to the update buffer).
+			fakeClient.InvokeWatchClusterCallback(test.newClusterUpdate, nil)
+			select {
+			case chu := <-ch.updateChannel:
+				if diff := cmp.Diff(chu.chu, []xdsclient.ClusterUpdate{test.newClusterUpdate}); diff != "" {
+					t.Fatalf("got unexpected cluster update, diff (-got, +want): %v", diff)
+				}
+			case <-ctx.Done():
+				t.Fatal("Timed out waiting for update from updateChannel.")
+			}
+		})
+	}
+}
+
+// TestUpdateRootClusterAggregateSuccess tests the case where an aggregate cluster is a root pointing to two child clusters
+// one of type EDS and the other of type LogicalDNS. This test will then send cluster updates for both the children, and
+// at the end there should be a successful clusterUpdate written to the update buffer to send back to CDS.
+func (s) TestUpdateRootClusterAggregateSuccess(t *testing.T) {
+	ch, fakeClient := setupTests(t)
+	ch.updateRootCluster(aggregateClusterService)
+
+	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer ctxCancel()
+	gotCluster, err := fakeClient.WaitForWatchCluster(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	}
+	if gotCluster != aggregateClusterService {
+		t.Fatalf("xdsClient.WatchCDS called for cluster: %v, want: %v", gotCluster, aggregateClusterService)
+	}
+	// The xdsClient telling the clusterNode that the cluster type is an aggregate cluster which will cause a lot of
+	// downstream behavior. For a cluster type that isn't an aggregate, the behavior is simple. The clusterNode will
+	// simply get a successful update, which will then ping the clusterHandler which will successfully build an update
+	// to send to the CDS policy. In the aggregate cluster case, the handleResp callback must also start watches for
+	// the aggregate cluster's children. The ping to the clusterHandler at the end of handleResp should be a no-op,
+	// as neither the EDS or LogicalDNS child clusters have received an update yet.
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType:             xdsclient.ClusterTypeAggregate,
+		ClusterName:             aggregateClusterService,
+		PrioritizedClusterNames: []string{edsService, logicalDNSService},
+	}, nil)
+
+	// xds client should be called to start a watch for one of the child clusters of the aggregate. The order of the children
+	// in the update written to the buffer to send to CDS matters, however there is no guarantee on the order it will start the
+	// watches of the children.
+	gotCluster, err = fakeClient.WaitForWatchCluster(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	}
+	if gotCluster != edsService {
+		if gotCluster != logicalDNSService {
```

Comment:
If's unfortunate that the order is now random.

Another way is `for { WaitForWatchCluster() }` until a timeout error, save the results in a slice, and compare the slice with the expected resources (`cmp.Diff()` has an option to sort the slice for you).

That way you can even have more than two child resources for the aggregated cluster.

Reply:
Eh, I don't like that idea. I think what's here is fine. The most important validation isn't this, but the ordering of the update passed to CDS Balancer, because in the end that's what really matters.